### PR TITLE
WebSurfer Documentation and Fixes

### PR DESCRIPTION
**Pull request by husseinmozannar**
This should not change how the WebSurfer operates in any way except for summarization/QA tool.

- adds documentation to web surfer
- adds documentation to playwright controller and exposes it
- cleans up code of web surfer
- adds markdown converter for summarization tool
- add on_message_stream with exposing inner messages
- adds model usage stats to web surfer
- add unit tests 

## Related issue number

Closes : 
#4573
#4567
#4565
#4467

### CODE REVIEW

**Comment by husseinmozannar** [comment]
Ready for review.

@afourney  this shouldn't change how the websurfer works, but added tests/docs/cleanup/markitdown for summarization. Take your time in review, this will probably be how the code looks like for a while.

@ekzhu  I added tests for playwright controller and websurfer. I had to check private attributes for the test and so plastered "# pyright: ignore[reportPrivateUsage]" everywhere, don't know how to otherwise. I do not want to expose these attributes otherwise

**Comment by husseinmozannar** [comment]
when reviewing, please take a look at how I changed the toml file to do playwright install, thanks!

File: python/packages/autogen-ext/src/autogen_ext/agents/web_surfer/playwright_controller.py [hunk 1]
```diff
@@ -1,9 +1,12 @@
 import asyncio
 import base64
+import io
 import os
 import random
 from typing import Any, Callable, Dict, Optional, Tuple, Union, cast
 
+# TODO: Fix unfollowed import
+from markitdown import MarkItDown  # type: ignore
```
Comment (afourney) [inline python/packages/autogen-ext/src/autogen_ext/agents/web_surfer/playwright_controller.py:9]: This is a pretty heavyweight dependency. I have mixed feelings about it including here, where it is used minimally (vs. other means of converting the DOM to text). Mabe we should make this optional (attempt the import in a try...catch), and backoff to something simpler if the dependency is not installed?

File: python/packages/autogen-ext/src/autogen_ext/agents/web_surfer/_multimodal_web_surfer.py [hunk 2]
```diff
@@ -59,67 +55,122 @@
...
+    MLM_HEIGHT = 765
+    MLM_WIDTH = 1224
+
+    SCREENSHOT_TOKENS = 1105
+
     def __init__(
         self,
         name: str,
         model_client: ChatCompletionClient,
+        downloads_folder: str | None = None,
```
Comment (ekzhu) [inline python/packages/autogen-ext/src/autogen_ext/agents/web_surfer/_multimodal_web_surfer.py:151]: At some point, I think we should have a separate interface/protocol for a web browser so we can enable other browser backends beside playwright.

File: python/packages/autogen-ext/src/autogen_ext/agents/web_surfer/_multimodal_web_surfer.py [hunk 3]
```diff
@@ -59,67 +55,122 @@
...
+    It must be used with a multimodal model client that supports tool calling, ideally GPT-4o currently.
 
+    .. note::
+        Please note that using the MultimodalWebSurfer involves interacting with a digital world designed for humans, which carries inherent risks.
+        Be aware that agents may occasionally attempt risky actions, such as recruiting humans for help or accepting cookie agreements without human involvement. Always ensure agents are monitored and operate within a controlled environment to prevent unintended consequences.
+        Moreover, be cautious that MultimodalWebSurfer may be susceptible to prompt injection attacks from webpages.
+
+    Args:
+        name (str): The name of the agent.
+        model_client (ChatCompletionClient): The model client used by the agent.
```
Comment (ekzhu) [inline python/packages/autogen-ext/src/autogen_ext/agents/web_surfer/_multimodal_web_surfer.py:76]: Add notes on the requirement on the model. 

And in the constructor validate model capabilities

File: python/packages/autogen-ext/src/autogen_ext/agents/web_surfer/_multimodal_web_surfer.py [hunk 4]
```diff
@@ -59,67 +55,122 @@
...
-VIEWPORT_WIDTH = 1440
+class MultimodalWebSurfer(BaseChatAgent):
+    """
+    MultimodalWebSurfer is a multimodal agent that acts as a web surfer that can search the web and visit web pages.
 
-# Size of the image we send to the MLM
-# Current values represent a 0.85 scaling to fit within the GPT-4v short-edge constraints (768px)
-MLM_HEIGHT = 765
-MLM_WIDTH = 1224
+    It launches a chromium browser and allows the playwright to interact with the web browser and can perform a variety of actions. The browser is launched on the first call to the agent and is reused for subsequent calls.
```
Comment (ekzhu) [inline python/packages/autogen-ext/src/autogen_ext/agents/web_surfer/_multimodal_web_surfer.py:65]: I think we can outline the steps taken by the agent in its on_messages_stream or on_messages method.

File: python/packages/autogen-ext/src/autogen_ext/agents/web_surfer/playwright_controller.py (hunk 1 above)
Comment (husseinmozannar) [reply python/packages/autogen-ext/src/autogen_ext/agents/web_surfer/playwright_controller.py]: added your suggestion, happy with both solutions

File: python/packages/autogen-ext/src/autogen_ext/agents/web_surfer/_multimodal_web_surfer.py (hunk 2 above)
Comment (husseinmozannar) [reply python/packages/autogen-ext/src/autogen_ext/agents/web_surfer/_multimodal_web_surfer.py:162]: yes we need to define other controller instances, but I don't think it's a priority for now

File: python/packages/autogen-ext/src/autogen_ext/agents/web_surfer/_multimodal_web_surfer.py (hunk 3 above)
Comment (husseinmozannar) [reply python/packages/autogen-ext/src/autogen_ext/agents/web_surfer/_multimodal_web_surfer.py]: done

File: python/packages/autogen-ext/src/autogen_ext/agents/web_surfer/_multimodal_web_surfer.py (hunk 4 above)
Comment (husseinmozannar) [reply python/packages/autogen-ext/src/autogen_ext/agents/web_surfer/_multimodal_web_surfer.py:65]: done

**Comment by husseinmozannar** [comment]
the test pass locally.. and passed before the latest commit while I didn't change any of the code functionally..

File: python/packages/autogen-ext/src/autogen_ext/agents/web_surfer/_multimodal_web_surfer.py (hunk 2 above)
Comment (afourney) [reply python/packages/autogen-ext/src/autogen_ext/agents/web_surfer/_multimodal_web_surfer.py:162]: I also think downloads folder is pretty universal. We had this even with the text-based browser before.

**Review by afourney — APPROVED** [review]
Looks good, as per the discussion.

File: python/packages/autogen-ext/src/autogen_ext/agents/web_surfer/_multimodal_web_surfer.py (hunk 2 above)
Comment (husseinmozannar) [reply python/packages/autogen-ext/src/autogen_ext/agents/web_surfer/_multimodal_web_surfer.py:162]: The reason to allow a None downloads folder is to allow option to disable downloads by default. I want the barebones websurfer to not require any creation of files or logs.

**Pull request merged**